### PR TITLE
[RyuJIT/ARM32] Kill the R4 register on exit if the call is the stub

### PR DESCRIPTION
**Pull request by sjsinju**
Related #11838

In this issue, farthestRefPhysRegRecord was nullptr in the end of allocateBusyReg(). That's because R4 register had been busy since the before allocation of R4 Register.
It was caused that the R4 register was not killed on exit of the stub call.

On ARM, There is the case that the argument of a virtual stub call is transferred by R4 register.
And then if the virtual stub is called, R4 register would be the busy status to avoid losing the argument data. It means R4 register must not be assigned to anywhere until the next kill.
So the R4 register has to be killed in case of after virtual stub call to escape the busy state.

## Before
```
---------------------------------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
Loc RP#  Name  Type  Action Reg  |r0   |r1   |r2   |r3   |r4   |r5   |r6   |r7   |r8   |r9   |r10  |r11  |r12  |sp   |lr   |
---------------------------------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
159.#169 V12   Use    Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
160.#170 r4    Fixd   Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
160.#171 I53   Def    PtArg r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
163.#172 V12   Use *  Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
164.#173 I54   Def    Alloc r3   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#174 r0    Fixd   Keep  r0   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#175 I52   Use *  Keep  r0   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#176 r4    Fixd   Keep  r4   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#177 I53   Use *  PtArg r4   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#178 I54   Use *  Keep  r3   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#179 r0    Kill   Keep  r0   |     |     |     |     |Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
.......
...
...
475.#443 BB15 PredBB13           |     |     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|V2  a|---- |---- |     |---- |     |
479.#444 r0    Fixd   Keep  r0   |     |     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|V2  a|---- |---- |     |---- |     |
479.#445 V2    Use *  Copy  r0   |V2  a|     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|V2  a|---- |---- |     |---- |     |
480.#446 r0    Fixd   Keep  r0   |     |     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|     |---- |---- |     |---- |     |
480.#447 I109  Def    Alloc r0   |I109a|     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|     |---- |---- |     |---- |     |
483.#448 r4    Fixd   Keep  r4   |I109a|     |V15 a|V14 a|Busy |V0  a|V3  a|V29 i|V25 i|     |---- |---- |     |---- |     |
483.#449 V15   Use<<<<<<<<<<<<<<<< Assertion
```

## After
```
---------------------------------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
Loc RP#  Name  Type  Action Reg  |r0   |r1   |r2   |r3   |r4   |r5   |r6   |r7   |r8   |r9   |r10  |r11  |r12  |sp   |lr   |
---------------------------------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
159.#171 r4    Fixd   Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
159.#172 V12   Use    Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
160.#173 r4    Fixd   Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
160.#174 I53   Def    PtArg r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
163.#175 V12   Use *  Keep  r4   |I52 a|     |     |     |V12 a|V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
164.#176 I54   Def    Alloc r3   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#177 r0    Fixd   Keep  r0   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#178 I52   Use *  Keep  r0   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#179 r4    Fixd   Keep  r4   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#180 I53   Use *  PtArg r4   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
165.#181 I54   Use *  Keep  r3   |I52 a|     |     |I54 a|Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#182 r0    Kill   Keep  r0   |     |     |     |     |Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#183 r1    Kill   Keep  r1   |     |     |     |     |Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#184 r2    Kill   Keep  r2   |     |     |     |     |Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#185 r3    Kill   Keep  r3   |     |     |     |     |Busy |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#186 r4    Kill   Keep  r4   |     |     |     |     |     |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#187 r12   Kill   Keep  r12  |     |     |     |     |     |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#188 lr    Kill   Keep  lr   |     |     |     |     |     |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
166.#189 r0    Fixd   Keep  r0   |     |     |     |     |     |V0  a|V1  i|V29 a|V25 a|V2  a|---- |---- |     |---- |     |
....
...
.
```

### CODE REVIEW

**Comment by hseok-oh** [comment]
There was a register name `REG_VIRTUAL_STUB_PARAM`. But It removed recently.
https://github.com/dotnet/coreclr/commit/8f8c5e8f3596c7e1607356894473db75b4353503

**Comment by sjsinju** [comment]
@hseok-oh OK. Thank you. Let me check.

**Comment by sjsinju** [comment]
@hseok-oh I've changed codes with reference to what you said. Please check it again.
/cc: @dotnet/arm32-contrib

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -2895,7 +2895,20 @@ regMaskTP LinearScan::getKillSetForNode(GenTree* tree)
                 }
                 else
                 {
-                    killMask = RBM_INT_CALLEE_TRASH;
+#ifdef _TARGET_ARM_
```
Comment (hseok-oh) [inline src/jit/lsra.cpp]: How about following code?
```cpp
killMask = RBM_INT_CALLEE_TRASH;

#ifdef _TARGET_ARM_
// comments...
if (tree->AsCall()->IsVirtualStub())
{
  regMaskTP virtualStubParamMask = ...
  killMask |= virtualStubParamMask;
}
#endif // _TARGET_ARM_
```

Comment (sjsinju) [inline src/jit/lsra.cpp]: @hseok-oh Looks good. Let me change codes. Thank you.

File: src/jit/lsra.cpp [hunk 2]
```diff
@@ -2896,6 +2896,18 @@ regMaskTP LinearScan::getKillSetForNode(GenTree* tree)
                 else
                 {
                     killMask = RBM_INT_CALLEE_TRASH;
+#ifdef _TARGET_ARM_
+                    // On ARM, There is the case that the argument of a virtual stub call is transferred by R4 register.
+                    // And then if the virtual stub is called,
+                    // R4 register would be the busy status to avoid losing the argument data.
+                    // It means R4 register must not be assigned to anywhere until the next kill.
+                    // But the R4 register has to be killed in case of after virtual stub call to escape the busy state.
```
Comment (BruceForstall) [inline src/jit/lsra.cpp]: This comment is quite confusing. @CarolEidt or @pgavlin Can you suggest a rewrite that would make it clearer?

File: src/jit/lsra.cpp [hunk 3]
```diff
@@ -2896,6 +2896,18 @@ regMaskTP LinearScan::getKillSetForNode(GenTree* tree)
...
+                    if (tree->AsCall()->IsVirtualStub())
+                    {
+                        regMaskTP virtualStubParamMask = compiler->virtualStubParamInfo->GetRegMask();
+                        killMask |= virtualStubParamMask;
```
Comment (BruceForstall) [inline src/jit/lsra.cpp]: This is ok, but IMO it's just as clear to use:
```
killMask |= compiler->virtualStubParamInfo->GetRegMask();
```

File: src/jit/lsra.cpp [hunk 4]
```diff
@@ -2896,6 +2896,18 @@ regMaskTP LinearScan::getKillSetForNode(GenTree* tree)
...
+                    }
+#endif // _TARGET_ARM_
```
Comment (BruceForstall) [inline src/jit/lsra.cpp]: This whole clause would be true when `compiler->compFloatingPointUsed` is true, also, so needs to move below the if/else.

**Comment by BruceForstall** [comment]
Here's a JitDump of the Linq test where this assert hits.

[dump.txt](https://github.com/dotnet/coreclr/files/1146628/dump.txt)

**Comment by CarolEidt** [comment]
> Also, why is this code ARM specific? Why is it that the virtual stub param register is trashed on ARM but no other platform?

This is because on other platforms the register is already in the RBM_INT_CALLEE_TRASH set. I suggested a comment that includes a reference to where the register is specified.

**Comment by BruceForstall** [comment]
> This is because on other platforms the register is already in the RBM_INT_CALLEE_TRASH set.

So we could do this for all platforms, then, to be correct, and also and avoid the `#ifdef`?

**Comment by CarolEidt** [comment]
> So we could do this for all platforms, then, to be correct, and also and avoid the #ifdef?

Yes, though I'm not sure the additional check would be optimized away, since the regMask is set in such an indirect fashion.

**Comment by BruceForstall** [comment]
Ok, then for performance (and this is a hot function), we maybe should have:
```
#ifndef _TARGET_ARM_
    // Verify that the special virtual stub call registers are in the kill mask. We don't just add them unconditionally to the killMask because for most architectures they are already in the RBM_CALLEE_TRASH set, and we don't want to introduce extra checks and calls in this hot function.
    assert(!tree->AsCall()->IsVirtualStub() || ((killMask & compiler->virtualStubParamInfo->GetRegMask()) == compiler->virtualStubParamInfo->GetRegMask()));
#endif // !_TARGET_ARM_
```

**Comment by CarolEidt** [comment]
> #ifndef _TARGET_ARM_
    // Verify that the special virtual stub call registers are in the kill mask. We don't just add them unconditionally to the killMask because for most architectures they are already in the RBM_CALLEE_TRASH set, and we don't want to introduce extra checks and calls in this hot function.
    assert(!tree->AsCall()->IsVirtualStub() || ((killMask & compiler->virtualStubParamInfo->GetRegMask()) == compiler->virtualStubParamInfo->GetRegMask()));
#endif // !_TARGET_ARM_

Sounds good, though once the relevant `_TARGET_ARM_` code is moved out of the if/then/else, perhaps this could just be an `#else` clause to that?

**Comment by sjsinju** [comment]
@BruceForstall  @CarolEidt Thank you guys for review, Let me check it.

**Comment by sjsinju** [comment]
@BruceForstall @CarolEidt I'v changed codes what you guys said. Could you check again this?

**Pull request merged**